### ISSUE TITLE
Slow code generation for log() on AMD CPUs

### ISSUE THREAD

**Issue by fehiepsi** [google/jax#5162]
I am using the latest version of jax/jaxlib. Currently, it took almost 90 seconds just to compute the following log of zeros.

```python
In [2]: import jax.numpy as jnp

In [3]: %time jnp.log(jnp.zeros((6, 2, 2, 294)))
CPU times: user 1min 27s, sys: 166 ms, total: 1min 27s
Wall time: 1min 27s
```

**Comment by hawkinsp** [comment]
I can't reproduce this.

Can you say a little more about:
* what exact versions of jax and jaxlib you have
* what hardware you are using: is this a CUDA or a non-CUDA build?
* what OS you are using
* what are your CPU and GPU models? (e.g., on Linux the output of `lscpu` and `nvidia-smi`, if applicable).

The worst timing I was able to get was ~7s for the first execution on CPU, after which the computation was cached and improved to a few milliseconds. That's slower than I'd like, but nowhere near as bad as you report.

**Comment by mattjj** [comment]
Wow, this is a funny behavior! I'm curious if we have any hypotheses for why this would be slow (to compile or execute) at all. Even Peter's numbers sound relatively slow.

**Comment by hawkinsp** [comment]
If I had to guess, won't this expression be constant-foldable by XLA? `jnp.zeros` will be lazy, and folded into the `log`, at which point XLA can use its (relatively slow) interpreter to constant-fold. But I'm really curious how that could take over a minute, though!

**Comment by mattjj** [comment]
Well, we don't fuse lazy values into computations; instead we force-materialize any computation arguments that are lazy values before dispatching the computation. #1960 was an implementation that did the fusion, but we decided not to merge it, I think mainly because we hadn't thought through the tradeoffs yet.

**Comment by fehiepsi** [comment]
So this is a regression in jaxlib 0.1.57. I installed several CPU versions and got

+ 0.1.57
```
In [2]: %time jax.numpy.log(jax.numpy.ones((6, 2, 2, 294)))
CPU times: user 1min 25s, sys: 68.2 ms, total: 1min 25s
Wall time: 1min 25s
```
+ 0.1.56
```
In [2]: %time jax.numpy.log(jax.numpy.ones((6, 2, 2, 294)))
CPU times: user 7.44 s, sys: 44.3 ms, total: 7.49 s
Wall time: 7.5 s
```

I guess @hawkinsp is using the unreleased jaxlib version 0.1.58, where this issue has been fixed (hopefully).

By the way, it is still a bit slow. It would be nice to improve the compiling time of numerical ops for multidimensional arrays. In [funsor](https://github.com/pyro-ppl/funsor/issues/393), most of the tests just run once (hence compiling time is expensive) and parameterized by different batch dimensions. We know that one source of slowness is the random samplers, and another source maybe this multidimensional one. Do you have any recommendations to improve testing speed? It is the most troublesome one when working with jax backend in `funsor`.

**Comment by hawkinsp** [comment]
Interesting. It appears to be an AMD code generation issue. I can reproduce on a GCP n2d VM, but not on any of the Intel machines I have access to.

**Comment by hawkinsp** [comment]
I still don't have a fix (I'm waiting for some of my more LLVM-adept colleagues to come back from vacation), but I did find that `XLA_FLAGS="--xla_backend_extra_options=--disable-post-ra"` seems to work around the issue.

**Comment by hawkinsp** [comment]
We bisected this to https://reviews.llvm.org/rG4b1120159274ca98b4a58064b116aba88372ac98 and https://reviews.llvm.org/D94182 will fix it. It will take a little while to get that into XLA, but we should be able to make a new jaxlib release once that is done.

**Comment by fehiepsi** [comment]
Awesome, thank you a lot, @hawkinsp!

**Comment by hawkinsp** [comment]
Jaxlib 0.1.58 should fix this. It's only built on Linux at the moment, but I don't think there are any Macs with AMD processors anyway...

**Comment by fehiepsi** [comment]
Thanks, Peter!
